dockerfiles/Dockerfile.cuda

```diff
@@ -2,40 +2,87 @@
+ARG CMAKE_CUDA_ARCHITECTURES="61;70;75;80;86;90"
 
-ENV	 DEBIAN_FRONTEND=noninteractive
+ENV DEBIAN_FRONTEND=noninteractive
 MAINTAINER Changming Sun "chasun@microsoft.com"
+
+# Add source code to /code
 ADD . /code
 
-ENV PATH /usr/local/nvidia/bin:/usr/local/cuda/bin:${PATH}
-RUN apt-get update \
-&& apt-get install -y --no-install-recommends ca-certificates g++ gcc make git python3-dev python3-pip
+ENV PATH=/usr/local/nvidia/bin:/usr/local/cuda/bin:${PATH}
+
+# Install required packages
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    ca-certificates \
+    g++ \
+    gcc \
+    make \
+    git \
+    cmake \
+    wget \
+    ninja-build \
+    python3-pip \
+    python3.12-dev \
+    python3.12-venv \
+    && rm -rf /var/lib/apt/lists/*
 
+# Install CUDNN 9.4.0.58 for building ONNX Runtime with CUDA.
+RUN wget https://developer.download.nvidia.com/compute/cudnn/redist/cudnn/linux-x86_64/cudnn-linux-x86_64-9.4.0.58_cuda12-archive.tar.xz \
+    && mkdir -p /code/build/cudnn \
+    && tar -Jxvf cudnn-linux-x86_64-9.4.0.58_cuda12-archive.tar.xz -C /code/build/cudnn --strip=1
+
+# Create a virtual environment and install dependencies, then build ONNX Runtime with CUDA support.
 RUN cd /code \
-&& python3 -m pip install --upgrade pip cmake psutil setuptools wheel packaging ninja \
-&& python3 -m pip install -r tools/ci_build/github/linux/docker/inference/x86_64/python/cpu/scripts/requirements.txt \
-&& /bin/bash ./build.sh --allow_running_as_root --skip_submodule_sync \
---use_cuda --cuda_home /usr/local/cuda --cudnn_home /usr/lib/x86_64-linux-gnu/ \
---build_shared_lib --skip_tests \
---config Release --build_wheel --update --build --parallel \
---cmake_generator Ninja \
---enable_cuda_nhwc_ops \
---cmake_extra_defines ONNXRUNTIME_VERSION=$(cat ./VERSION_NUMBER) \
-onnxruntime_BUILD_UNIT_TESTS=OFF \
-`CMAKE_CUDA_ARCHITECTURES=${CMAKE_CUDA_ARCHITECTURES}`
-
-FROM nvcr.io/nvidia/cuda:12.5.1-cudnn-runtime-ubuntu22.04
-ENV	 DEBIAN_FRONTEND=noninteractive
-COPY --from=0 /code/build/Linux/Release/dist /root
+    && python3 -m venv /code/env \
+    && . /code/env/bin/activate \
+    && pip install --upgrade psutil setuptools wheel packaging \
+    && pip install -r tools/ci_build/github/linux/docker/inference/x86_64/python/cpu/scripts/requirements.txt \
+    && python /code/tools/ci_build/build.py --build_dir /code/build/Linux \
+    --allow_running_as_root --skip_submodule_sync \
+    --use_cuda --cuda_home /usr/local/cuda \
+    --cudnn_home /code/build/cudnn \
+    --build_shared_lib --skip_tests \
+    --config Release --build_wheel --update --build --parallel \
+    --cmake_generator Ninja \
+    --enable_cuda_nhwc_ops \
+    --cmake_extra_defines ONNXRUNTIME_VERSION=$(cat ./VERSION_NUMBER) "CMAKE_CUDA_ARCHITECTURES=${CMAKE_CUDA_ARCHITECTURES}" onnxruntime_BUILD_UNIT_TESTS=OFF
+
+# Start second stage to copy the build artifacts
+FROM nvcr.io/nvidia/cuda:12.6.1-runtime-ubuntu24.04
```

Comment:
So, the image doesn't have cudnn? Why don't you choose one with cudnn?

Reply:
Just to demonstrate that we can use `pip install nvidia-cudnn-cu12` to install cudnn for python.